### PR TITLE
Adds recall@k metric to rank eval API

### PR DESCRIPTION
**Pull request by joshdevins**
This change adds the recall@k metric and refactors precision@k to match the new metric.

Recall@k is an important metric to use for learning to rank (LTR). Candidate generation / first phase ranking functions are often optimized for high recall, in order to generate as many relevant candidates in the top-k as possible for a second phase of ranking using LTR or a less efficient ranking function. Adding this metric allows tuning the candidate generation for LTR.

See: https://github.com/elastic/elasticsearch/issues/51676

### CODE REVIEW

**Comment by joshdevins** [comment]
I'll be adding some more tests but I wanted to create a draft PR for review once I had something worth sharing 😉

**Review by jtibshirani — COMMENTED** [review]
I did an initial pass, but haven't yet done a detailed review as it sounds like you're still sorting out some details.

For me, the new confusion matrix abstraction adds more complexity than it’s worth:
* We compute true negatives, but none of our metrics actually use that value yet. To calculate true negatives, we must add a new 'total hits' parameter to `EvaluationMetric#evaluate`.
* To me it doesn’t make it easier to understand each calculation (although maybe others have a different intuition).

Perhaps we could hold off on adding an abstraction until we add more metrics that would benefit from sharing logic?

**Comment by joshdevins** [comment]
> We compute true negatives, but none of our metrics actually use that value yet
Yeah, I stuck it in to see how it would look, but I agree it's more than we need and can take it out.

> To me it doesn’t make it easier to understand each calculation
That's basically why I added it — I find it easier to reason about these calculations with a confusion matrix, thinking about true positives, false positives, etc. instead of the more abstract concepts like recall being "the proportion of relevant documents in the top-k vs all possible relevant documents", but maybe that's just me 😄 I'm not very opinionated about that detail so I'm happy either way.

**Comment by joshdevins** [comment]
> We compute true negatives, but none of our metrics actually use that value yet
Yeah, I stuck it in to see how it would look, but I agree it's more than we need and can take it out.

> To me it doesn’t make it easier to understand each calculation
That's basically why I added it — I find it easier to reason about these calculations with a confusion matrix, thinking about true positives, false positives, etc. instead of the more abstract concepts like recall being "the proportion of relevant documents in the top-k vs all possible relevant documents", but maybe that's just me 😄 I'm not very opinionated about that detail so I'm happy either way.

**Comment by joshdevins** [comment]
@elasticmachine run elasticsearch-ci

**Comment by joshdevins** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by joshdevins** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by joshdevins** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by joshdevins** [comment]
zOMG, builds are finally passing. I'm removing the confusion matrix abstraction, adding some better recall tests, then I'll ping for a review again after that. @sbourke this should make adding MAP a bit easier, so you can base it off this change set.

**Comment by joshdevins** [comment]
I'm adding Docs as well. I just realised I can do that in the PR. Will get to that tomorrow though.

**Review by jtibshirani — COMMENTED** [review]
This is looking good to me, just left a few small comments. Some higher-level notes:
* There is quite a bit of duplication between `PrecisionAtK` and `RecallAtK` in terms of serialization code, getters + setters, etc. We could try a refactor to pull out some shared code, but to me it's best to keep the implementation straightforward for now.
* I think the current approach is fine from a backwards-compatibility perspective. When nodes containing this new metric try to communicate with nodes without the implementation, we will simply fail (because we won't find a matching 'named writeable').

**Comment by joshdevins** [comment]
> There is quite a bit of duplication between PrecisionAtK and RecallAtK in terms of serialization code, getters + setters, etc. We could try a refactor to pull out some shared code, but to me it's best to keep the implementation straightforward for now.

Yeah, that's why I tried to factor it out in the confusion matrix version. The serialization stuff is annoying to factor out since it relies on a lot of static methods and fields. I would vote to leave it as-is for now and perhaps after we add MAP/GMAP we start to see better patterns to refactor to. In general, I would prefer to group these metrics into top-k set-based metrics and rank metrics. I think we can find good ways to factor things based on slightly different abstractions for these metrics. The confusion matrix metric was a first attempt to do that. It's also how we have done these metrics in the ML. For classification, those are based on a [confusion matrix metric](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/softclassification/ConfusionMatrix.java).

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RecallAtK.java
Comment (joshdevins) [reply modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RecallAtK.java]: Copy-paste error. I fixed this in `PrecisionAtK` as well.

File: modules/rank-eval/src/test/java/org/elasticsearch/index/rankeval/PrecisionAtKTests.java
Comment (joshdevins) [reply modules/rank-eval/src/test/java/org/elasticsearch/index/rankeval/PrecisionAtKTests.java]: Yeah, I was trying to fix the test names in `PrecisionAtK` which I found contradictory between the test/method name and the details of the test. I added the `AtFive` when the test was specifically not using the default `k` since it better describes the test case. I'm happy to drop all those though and simplify the names.

File: modules/rank-eval/src/test/java/org/elasticsearch/index/rankeval/RankEvalSpecTests.java
Comment (joshdevins) [reply modules/rank-eval/src/test/java/org/elasticsearch/index/rankeval/RankEvalSpecTests.java]: Oops, missed this. Thanks! I think I did it correctly, check the next commit and let me know.

**Review by jtibshirani — APPROVED** [review]
> I would vote to leave it as-is for now and perhaps after we add MAP/GMAP we start to see better patterns to refactor to.

Sounds good!

This looks good to me, I think it's ready to merge after you add a note to the docs. If you agree, it'd be good to remove the 'WIP' label, and add the appropriate labels (area, type of change, plus target versions).

File: docs/reference/search/rank-eval.asciidoc
Comment (jtibshirani) [inline docs/reference/search/rank-eval.asciidoc]: `precision` -> `recall`

Comment (joshdevins) [reply docs/reference/search/rank-eval.asciidoc]: 😭

**Pull request merged**
